File: leetcode/april2024/13_maximal_rectangle.py
# CODE

class Solution:
    def largestRectangleArea(self, histogram):
        stack = deque()
        area = 0
        histogram.append(0)  # Append a sentinel to handle the end condition
        for i, height in enumerate(histogram):
            while stack and height <= histogram[stack[-1]]:
                h = histogram[stack.pop()]
                w = i if not stack else i - stack[-1] - 1
                area = max(area, h * w)
            stack.append(i)
        return area

    def maximalRectangle(self, matrix):
        maxArea = 0
        prefixHeight = [0] * len(matrix[0]) if matrix else []
        for row in matrix:
            for j, cell in enumerate(row):
                prefixHeight[j] = prefixHeight[j] + 1 if cell == '1' else 0
            maxArea = max(maxArea, self.largestRectangleArea(prefixHeight))
        return maxArea

File: leetcode/2024_4_april/c++/13_maximal_rectangle.cpp
// CODE

class Solution {
    public:
    int largestRectangleArea(vector<int> &histogram) {
        stack<int> s;
        int area = 0, n = histogram.size();
        for(int i = 0; i <= n; i++) {
            while(!s.empty() and (i == n or histogram[s.top()] >= histogram[i])) {
                int w, h = histogram[s.top()];
                s.pop();
                if(s.empty()) w = i;
                else w = i - s.top() - 1;
                area = max(area, h * w);
            }
            s.push(i);
        }
        return area;
    }

    int maximalRectangle(vector<vector<char>>& matrix) {
        int maxArea = 0, m = matrix.size(), n = matrix[0].size();
        vector<int> prefixHeight(n, 0);
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(matrix[i][j] == '1') prefixHeight[j]++;
                else prefixHeight[j] = 0;
            }
            int area = largestRectangleArea(prefixHeight);
            maxArea = max(maxArea, area);
        }
        return maxArea;
    }
};